src/custom/constants/networks.ts

```diff
@@ -1,30 +1,146 @@
-  // [SupportedChainId.OPTIMISTIC_KOVAN]: `https://optimism-kovan.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.ARBITRUM_ONE]: `https://arbitrum-mainnet.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.ARBITRUM_RINKEBY]: `https://arbitrum-rinkeby.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.POLYGON]: `https://polygon-mainnet.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.POLYGON_MUMBAI]: `https://polygon-mumbai.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.CELO]: `https://forno.celo.org`,
-  // [SupportedChainId.CELO_ALFAJORES]: `https://alfajores-forno.celo-testnet.org`,
-  [SupportedChainId.GNOSIS_CHAIN]: `https://rpc.gnosis.gateway.fm`,
+class BaseUrlJsonRpcProvider extends UrlJsonRpcProvider {}
+type ProviderClass = Newable<typeof BaseUrlJsonRpcProvider>
+
+// List of public providers that don't require an API key, per chain support.
+const PUBLIC_PROVIDER_LIST: Record<SupportedChainId, ProviderClass[]> = {
+  [SupportedChainId.MAINNET]: [InfuraProvider, CloudflareProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GOERLI]: [InfuraProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GNOSIS_CHAIN]: [],
+}
+
+function getProvider(chainId: SupportedChainId): JsonRpcProvider[] {
+  const result: JsonRpcProvider[] = []
+
+  if (window.ethereum) {
+    result.push(new Web3Provider(window.ethereum, RPC_CONFIG[chainId]))
+  }
+
+  if (typeof RPC_CONFIG[chainId].rpcUrl === 'string') {
+    result.push(new StaticJsonRpcProvider(RPC_CONFIG[chainId].rpcUrl))
+  } else {
+    result.push(new InfuraProvider(RPC_CONFIG[chainId], INFURA_KEY))
+  }
+
+  if (result.length > 0) {
+    return result
+  }
+
+  for (const Provider of PUBLIC_PROVIDER_LIST[chainId]) {
+    try {
+      const provider = new Provider(RPC_CONFIG[chainId])
+
+      result.push(provider)
+    } catch (error) {
+      const { sentryError, tags } = constructSentryError(error, { message: "Couldn't create public provider" })
+
+      Sentry.captureException(sentryError, { tags })
+    }
+  }
+
+  return result
+}
+
+export const PROVIDERS: Record<SupportedChainId, JsonRpcProvider[]> = {
+  [SupportedChainId.MAINNET]: getProvider(SupportedChainId.MAINNET),
+  [SupportedChainId.GOERLI]: getProvider(SupportedChainId.GOERLI),
+  [SupportedChainId.GNOSIS_CHAIN]: getProvider(SupportedChainId.GNOSIS_CHAIN),
+}
+
+export const getRpcUrls = (chainId: SupportedChainId) => PROVIDERS[chainId].map((provider) => provider.connection.url)
+
+export const PROVIDERS_RPC_URLS: Record<SupportedChainId, string[]> = {
+  [SupportedChainId.MAINNET]: getRpcUrls(SupportedChainId.MAINNET),
+  [SupportedChainId.GOERLI]: getRpcUrls(SupportedChainId.GOERLI),
+  [SupportedChainId.GNOSIS_CHAIN]: getRpcUrls(SupportedChainId.GNOSIS_CHAIN),
+}
+
+export const PROVIDERS_RPC_URLS_FIRST_ONLY: Record<SupportedChainId, string> = {
+  [SupportedChainId.MAINNET]: PROVIDERS_RPC_URLS[SupportedChainId.MAINNET][0],
```

Comment:
nit: this `PROVIDERS` and `PROVIDERS_RPC_URLS` is a bit repetitive. Considere using an IIFE returning the 3 consts, the function would just `reduce` the list of supported chains

Reply:
I don't think we need this, right? we already have `PROVIDERS`?